tensorflow_addons/custom_ops/image/python/dense_image_warp_test.py

```diff
@@ -0,0 +1,256 @@
+        atol = 1e-6
+        rtol = 1e-6
+        if low_precision:
+            atol = 1e-2
+            rtol = 1e-3
+        self.assertAllClose(
+            interp,
+            pred_interpolation[batch_index, y_index, x_index, :],
+            atol=atol,
+            rtol=rtol)
+
+    def check_zero_flow_correctness(self, shape, image_type, flow_type):
+        """Assert using zero flows doesn't change the input image."""
+
+        rand_image, rand_flows = self.get_random_image_and_flows(
+            shape, image_type, flow_type)
+        rand_flows *= 0
+
+        interp = dense_image_warp_ops.dense_image_warp(
+            image=tf.convert_to_tensor(rand_image),
+            flow=tf.convert_to_tensor(rand_flows))
+
+        self.assertAllClose(rand_image, interp)
+
+    @tf_test_util.run_all_in_graph_and_eager_modes
+    def test_zero_flows(self):
+        """Apply check_zero_flow_correctness() for a few sizes and types."""
+
+        shapes_to_try = [[3, 4, 5, 6], [1, 2, 2, 1]]
+        for shape in shapes_to_try:
+            self.check_zero_flow_correctness(
+                shape, image_type="float32", flow_type="float32")
+
+    def check_interpolation_correctness(self,
+                                        shape,
+                                        image_type,
+                                        flow_type,
+                                        num_probes=5):
+        """Interpolate, and then assert correctness for a few query locations."""
+        low_precision = image_type == "float16" or flow_type == "float16"
+        rand_image, rand_flows = self.get_random_image_and_flows(
+            shape, image_type, flow_type)
+
+        interp = dense_image_warp_ops.dense_image_warp(
+            image=tf.convert_to_tensor(rand_image),
+            flow=tf.convert_to_tensor(rand_flows))
+
+        for _ in range(num_probes):
+            batch_index = np.random.randint(0, shape[0])
+            y_index = np.random.randint(0, shape[1])
+            x_index = np.random.randint(0, shape[2])
+
+            self.assert_correct_interpolation_value(
+                rand_image,
+                rand_flows,
+                interp,
+                batch_index,
+                y_index,
+                x_index,
+                low_precision=low_precision)
+
+    @tf_test_util.run_all_in_graph_and_eager_modes
+    def test_interpolation(self):
+        """Apply check_interpolation_correctness() for a few sizes and types."""
+
+        shapes_to_try = [[3, 4, 5, 6], [1, 5, 5, 3], [1, 2, 2, 1]]
+        for im_type in ["float32", "float64", "float16"]:
+            for flow_type in ["float32", "float64", "float16"]:
+                for shape in shapes_to_try:
+                    self.check_interpolation_correctness(
+                        shape, im_type, flow_type)
+
+    # TODO: switch to TF2 later.
+    @tf_test_util.run_deprecated_v1
+    def test_gradients_exist(self):
```

Comment:
Wondering if there is a more elegant way to test gradients in both eager and graph modes.

Reply:
Don't worry. We don't find a workaround now, and will clean it in the future.